File: Trees/Subtree of Another Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) return true;
        else if (root != null && subRoot == null) return true;
        else if (root == null && subRoot != null) return false;
        
        if (root.val == subRoot.val) 
        {
            if (IsSameTree(root,subRoot)) return true;
        }
        
        //traverse node until value matching subRoot value is found
        if (IsSubtree(root.left, subRoot) || IsSubtree(root.right, subRoot)) 
            return true;
        
        return false;
    }
    
    public bool IsSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        else if ((p != null && q == null) || (p == null && q != null)) return false;

        if (p.val != q.val) return false;
        
        if (!IsSameTree(p.left, q.left) || !IsSameTree(p.right,q.right)) return false;
        
        return true;
    }
    
}

File: Trees/Subtree of Another Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
    if (!subRoot) return true;
    if (!root) return false;

    if (isSameTree(root, subRoot)) return true;

    return (isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot));
    
    function isSameTree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
        if (!root && !subRoot) return true;
        else if (root == null || subRoot == null) return false;
        else
        {
            if (root.val != subRoot.val) return false;
            const isLeftEqual = isSameTree(root.left, subRoot.left);
            const isRightEqual = isSameTree(root.right, subRoot.right);

            return isLeftEqual && isRightEqual;
        } 
    }
};
